// @farbenmeer/react-spring-slider/dist/index.stories.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importStar(require("react"));
var react_2 = require("@storybook/react");
var _1 = __importDefault(require("."));
var images = [
    "https://images.pexels.com/photos/296878/pexels-photo-296878.jpeg?auto=compress&cs=tinysrgb&dpr=2&h=750&w=1260",
    "https://images.pexels.com/photos/1509428/pexels-photo-1509428.jpeg?auto=compress&cs=tinysrgb&dpr=2&h=750&w=1260",
    "https://images.pexels.com/photos/351265/pexels-photo-351265.jpeg?auto=compress&cs=tinysrgb&dpr=2&h=750&w=1260",
    "https://images.pexels.com/photos/924675/pexels-photo-924675.jpeg?auto=compress&cs=tinysrgb&dpr=2&h=750&w=1260"
];
var onSlideChange = function (index) {
    console.log("Slide changed to: " + index);
};
var imageStyle = function (src) { return ({
    backgroundSize: "cover",
    backgroundImage: "url(" + src + ")",
    height: "100%",
    width: "100%"
}); };
react_2.storiesOf("Slider", module)
    .add("Default", function () { return (react_1.default.createElement("div", { style: { width: "100vw", height: "100vh" } },
    react_1.default.createElement(_1.default, { hasBullets: true, onSlideChange: onSlideChange }, images.map(function (image) { return (react_1.default.createElement("div", { key: image, draggable: "false", style: imageStyle(image) })); })))); })
    .add("Smaller than 100%", function () { return (react_1.default.createElement("div", { style: { width: "500px", height: "500px" } },
    react_1.default.createElement(_1.default, { hasBullets: true }, images.map(function (image) { return (react_1.default.createElement("div", { key: image, draggable: "false", style: imageStyle(image) })); })))); })
    .add("With auto sliding", function () { return (react_1.default.createElement("div", { style: { width: "100vw", height: "100vh" } },
    react_1.default.createElement(_1.default, { hasBullets: true, auto: 2000 }, images.map(function (image) { return (react_1.default.createElement("div", { key: image, draggable: "false", style: imageStyle(image) })); })))); })
    .add("With activeIndex (2)", function () { return (react_1.default.createElement("div", { style: { width: "100vw", height: "100vh" } },
    react_1.default.createElement(_1.default, { hasBullets: true, activeIndex: 2 }, images.map(function (image) { return (react_1.default.createElement("div", { key: image, draggable: "false", style: imageStyle(image) })); })))); })
    .add("With activeIndex (interval)", function () {
    var _a = react_1.useState(2), activeIndex = _a[0], setActiveIndex = _a[1];
    react_1.useEffect(function () {
        var id = setInterval(function () {
            var index = activeIndex + (1 % images.length);
            setActiveIndex(index);
        }, 2000);
        return function () {
            clearInterval(id);
        };
    });
    return (react_1.default.createElement("div", { style: { width: "100vw", height: "100vh" } },
        react_1.default.createElement(_1.default, { hasBullets: true, activeIndex: activeIndex }, images.map(function (image) { return (react_1.default.createElement("div", { key: image, draggable: "false", style: imageStyle(image) })); }))));
})
    .add("With custom setSlide function", function () {
    var _a = react_1.useState(0), activeIndex = _a[0], setActiveIndex = _a[1];
    var setSlideCustom = function (slide) { return 1; };
    return (react_1.default.createElement("div", { style: { width: "100vw", height: "100vh" } },
        react_1.default.createElement(_1.default, { hasBullets: true, activeIndex: activeIndex, setSlideCustom: setSlideCustom }, images.map(function (image) { return (react_1.default.createElement("div", { key: image, draggable: "false", style: imageStyle(image) })); }))));
})
    .add("With different `bulletStyle`", function () {
    return (react_1.default.createElement("div", { style: { width: "100vw", height: "100vh" } },
        react_1.default.createElement(_1.default, { hasBullets: true, bulletStyle: { backgroundColor: "#fff" } }, images.map(function (image) { return (react_1.default.createElement("div", { key: image, draggable: "false", style: imageStyle(image) })); }))));
})
    .add("With custom `bullets`", function () {
    var BulletComponent = function (_a) {
        var onClick = _a.onClick, isActive = _a.isActive;
        return (react_1.default.createElement("li", { style: {
                width: "25px",
                height: "25px",
                backgroundColor: "red",
                margin: "0 2px",
                opacity: isActive ? "0.5" : undefined
            }, onClick: onClick }));
    };
    return (react_1.default.createElement("div", { style: { width: "100vw", height: "100vh" } },
        react_1.default.createElement(_1.default, { hasBullets: true, BulletComponent: BulletComponent }, images.map(function (image) { return (react_1.default.createElement("div", { key: image, draggable: "false", style: imageStyle(image) })); }))));
})
    .add("With arrows", function () {
    return (react_1.default.createElement("div", { style: { width: "100vw", height: "100vh" } },
        react_1.default.createElement(_1.default, { hasArrows: true }, images.map(function (image) { return (react_1.default.createElement("div", { key: image, draggable: "false", style: imageStyle(image) })); }))));
})
    .add("With custom arrow style", function () {
    var arrowStyle = { border: "solid red", borderWidth: "0 5px 5px 0" };
    return (react_1.default.createElement("div", { style: { width: "100vw", height: "100vh" } },
        react_1.default.createElement(_1.default, { hasArrows: true, arrowStyle: arrowStyle }, images.map(function (image) { return (react_1.default.createElement("div", { key: image, draggable: "false", style: imageStyle(image) })); }))));
})
    .add("With custom arrow component", function () {
    var ArrowComponent = function (_a) {
        var onClick = _a.onClick, direction = _a.direction;
        return (react_1.default.createElement("div", { style: {
                border: "1px solid black",
                padding: "1em",
                backgroundColor: "white"
            }, onClick: onClick }, direction));
    };
    return (react_1.default.createElement("div", { style: { width: "100vw", height: "100vh" } },
        react_1.default.createElement(_1.default, { hasArrows: true, ArrowComponent: ArrowComponent }, images.map(function (image) { return (react_1.default.createElement("div", { key: image, draggable: "false", style: imageStyle(image) })); }))));
})
    .add("With arrows and bullets", function () {
    return (react_1.default.createElement("div", { style: { width: "100vw", height: "100vh" } },
        react_1.default.createElement(_1.default, { hasArrows: true, hasBullets: true }, images.map(function (image) { return (react_1.default.createElement("div", { key: image, draggable: "false", style: imageStyle(image) })); }))));
})
    .add("With onClick", function () { return (react_1.default.createElement("div", { style: { width: "100vw", height: "100vh" } },
    react_1.default.createElement(_1.default, { hasBullets: true, onSlideChange: onSlideChange }, images.map(function (image) { return (react_1.default.createElement("div", { key: image, draggable: "false", style: imageStyle(image), onClick: function () { return console.log("click"); } })); })))); });
